File: client/java/src/main/java/com/be/mesh/client/prsim/Publisher.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.prsim;

import com.be.mesh.client.annotate.Index;
import com.be.mesh.client.annotate.MPI;
import com.be.mesh.client.annotate.SPI;
import com.be.mesh.client.struct.Event;
import com.be.mesh.client.struct.Principal;
import com.be.mesh.client.struct.Topic;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Event publisher.
 *
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public interface Publisher {

    /**
     * Publish event to mesh.
     *
     * @param events events
     * @return results
     */
    @MPI("mesh.queue.publish")
    List<String> publish(@Index(0) List<Event> events);

    /**
     * Synchronized broadcast the event to all subscriber. This maybe timeout with to many subscriber.
     *
     * @param events Event payload
     * @return Synchronized subscriber return value
     */
    @MPI("mesh.queue.broadcast")
    List<String> broadcast(@Index(0) List<Event> events);

    /**
     * Publish message to local node.
     *
     * @param binding topic
     * @param payload structure
     * @return eventId
     */
    default String publish(Topic binding, Object payload) {
        Event event = Event.newInstance(payload, binding);
        return this.publish(Collections.singletonList(event)).get(0);
    }

    /**
     * Unicast will publish to another node.
     *
     * @param binding topic
     * @param payload payload
     * @return eventId
     */
    default String unicast(Topic binding, Object payload, Principal principal) {
        Event event = Event.newInstance(payload, binding, principal);
        return this.publish(Collections.singletonList(event)).get(0);
    }

    /**
     * Multicast will publish event to principal groups.
     *
     * @param binding    topic.
     * @param payload    payload
     * @param principals principal group.
     * @return event id
     */
    default List<String> multicast(Topic binding, Object payload, List<Principal> principals) {
        List<Event> events = principals.stream().map(principal -> Event.newInstance(payload, binding, principal)).collect(Collectors.toList());
        return this.publish(events);
    }

    /**
     * Synchronized broadcast the event to all subscriber. This maybe timeout with to many subscriber.
     *
     * @param payload Event payload
     * @return Synchronized subscriber return value
     */
    default List<String> broadcast(Topic binding, Object payload) {
        return this.broadcast(Collections.singletonList(Event.newInstance(payload, binding)));
    }
}

File: client/python/mesh/prsim/publisher.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
from abc import ABC, abstractmethod
from typing import List, Any

from mesh.kinds.event import Event
from mesh.kinds.event import Topic
from mesh.kinds.principal import Principal
from mesh.macro import mpi, spi


@spi("mesh")
class Publisher(ABC):

    @abstractmethod
    @mpi("mesh.queue.publish")
    def publish(self, events: List[Event]) -> List[str]:
        """
        Publish event to mesh.
        :param events:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.queue.broadcast")
    def broadcast(self, events: List[Event]) -> List[str]:
        """
        Synchronized broadcast the event to all subscriber. This maybe timeout with to many subscriber.
        :param events: Event payload
        :return: Synchronized subscriber return value
        """
        pass

    def publish_with_topic(self, binding: Topic, payload: Any) -> str:
        """
        Publish message to local node.
        :param binding:
        :param payload:
        :return:
        """
        event = Event.new_instance(payload, binding)
        return self.publish([event])[0]

    def unicast(self, binding: Topic, payload: Any, principal: Principal) -> str:
        """
        Unicast will publish to another node.
        :param binding:
        :param payload:
        :param principal:
        :return:
        """
        event = Event.new_instance_with_target(payload, binding, principal)
        return self.publish([event])[0]

    def multicast(self, binding: Topic, payload: Any, principals: List[Principal]) -> List[str]:
        """
        Multicast will publish event to principal groups.
        :param binding:
        :param payload:
        :param principals:
        :return:
        """
        events = []
        for principal in principals:
            events.append(Event.new_instance_with_target(payload, binding, principal))
        return self.publish(events)

    def broadcast_with_topic(self, binding: Topic, payload: Any) -> List[str]:
        """
        Synchronized broadcast the event to all subscriber. This maybe timeout with to many subscriber.
        :param binding:
        :param payload:
        :return:
        """
        return self.broadcast([Event.new_instance(payload, binding)])
